### ISSUE TITLE
Introduce factory for AdSense response mock data

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#2585]
After #2235, the `module-adsense-components.stories.js` file is now full of inlined fixtures, which is due to the mock data being dependent on the `referenceDate` used for the story. While we could have continued to use fixtures, doing so is in another way confusing as it decouples the two pieces entirely and obfuscates that they are connected.

We should introduce an AdSense-specific test utility, which generates AdSense response data.

We should also think about how this would affect request args definition in Storybook and tests - whether we should use fixtures there, factories, or just inline them (since those are _a lot_ simpler than response data).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* There should be a function like `getAdSenseMockResponse( requestArgs )` which generates a full AdSense response for the given request `requestArgs`.
    * This can basically replicate logic which we already have in PHP via the Site Kit tester plugin. It'll be based on some hard-coded mock values, combined with logic to arrange them accordingly to get the correct response format for the request. It should also follow any conventions introduced in #2584.
* The AdSense stories should use the above function instead of inlining response mock data.
* All AdSense fixtures for report data or args should be deleted from the plugin.

## Implementation Brief

* Add `getAdSenseMockResponse( requestArgs )` function in a new file `modules/adsense/util/data-mock.js`. This function is responsible for generating mock responses and should use the similar approach from #2584. It should also use `rxjs` to generate and process response data.
  * Generate an args hash and use it as a new seed value for the faker as we do it in the data mock function in the Analytics module (see [here](https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/analytics/util/data-mock.js#L176-L187)). But for the AdSense module, let's generate a hash using JSON stringified version of the `requestArgs` variable.
  * Define a new object `data` that replicates the response data object.
  * Create a new rxjs stream with dates starting from `startDate` and ending with `endDate` if the `requestArgs` contains the `DATE` dimension (see an example [here](https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/analytics/util/data-mock.js#L215-L226)). If the `DATE` dimension is not selected, then create a new stream with a single `null` value like it's done [here](https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/analytics/util/data-mock.js#L245-L246).
  * Pipe following rxjs operators to the newly created stream using [`.pipe()`](https://rxjs-dev.firebaseapp.com/guide/operators) method:
    * The first [`map`](https://rxjs-dev.firebaseapp.com/api/operators/map) operator to convert stream elements to an array. If an element value is not empty, then it should be added to the returning array. Otherwise, the `map` operation should just return an empty array.
    
      _At this step, a stream of dates will be converted to a stream of rows (an array of arrays) with dates if `DATE` dimension has been requested (`--- [ '2020-09-06' ] ---> [ '2020-09-07' ] ---> ... `), otherwise just a stream of a single empty row._
    * The second `map` operator that should add metric values to the incoming array (a row) and return it. Use the `faker` library to generate random values and respect metric types to generate correct values. As to the metrics, we use the following metrics and need to support them in this function: `EARNINGS`, `PAGE_VIEWS_RPM`, `IMPRESSIONS`, and `PAGE_VIEWS_CTR`. Check the documentation for more details: https://developers.google.com/adsense/management/metrics-dimensions
    
      _By doing it, we will get a stream with fully provisioned rows: `--- [ '2020-09-06', metric1ValueA, metric2ValueA, ... ] ---> [ '2020-09-07', metric1ValueB, metric2ValueB, ... ] ---> ...` if `DATE` dimension has been requested, or just a single row with metric values `--- [ metric1Value, metric2Value ] ---|->`._
    * The [`reduce`](https://rxjs-dev.firebaseapp.com/api/operators/reduce) operator to combine all rows in the stream into a single item (an array of rows). See an example: https://github.com/google/site-kit-wp/blob/699ca8d5ba5f4fe0c9e96fee2075826806bb418d/assets/js/modules/analytics/util/data-mock.js#L259-L260
      _Now the stream should contain all rows as a single item: `--- [ [ ... ], [ ... ], ... ] ---|->`._
  * Add a subscribe callback to the stream using [`.subscribe()`](https://rxjs-dev.firebaseapp.com/guide/subscription) method. This callback will retrieve the resulting array of rows, assign it to the `data.rows` property.
  * Populate other properties of the `data` object like `totals` and `averages` using information from the rows array.
  * Return the `data` object.
* Update stories in the `stories/module-adsense-components.stories.js` file to use the new `getAdSenseMockResponse` function instead of hardcoded fixtures.
* Find all usages of fixtures from `modules/adsense/datastore/__fixtures__/earnings.json` and `modules/adsense/datastore/__fixtures__/report.json` files and rework them to use he new `getAdSenseMockResponse` function.
* Delete `modules/adsense/datastore/__fixtures__/earnings.json` and `modules/adsense/datastore/__fixtures__/report.json` files.

### Test Coverage

* N/A

### Visual Regression Changes

* N/A

## QA Brief

* Verify that storybook stories for AdSense widgets work correctly and display correct graphs.

## Changelog entry

* N/A

**Comment by felixarntz** [comment]
IB ✅ 

Regarding Storybook, there is some overlap with the work done in #2065 here. Due to the other conversation about how to implement the Storybook coverage in that issue, it might make sense to completely hold off and have a follow-up issue for only the Storybook story which can then depend on this one, to avoid unnecessary extra work. Since Storybook coverage wasn't explicitly requested in the ACs, I think in this case it would be okay to defer (although we should add it in general). cc @aaemnnosttv @asvinb

**Comment by eugene-manuilov** [comment]
@aaemnnosttv this is ready for your review. I have updated the generated data to match the format used by API v2, so we should be good to go if you no objections.

**Comment by aaemnnosttv** [comment]
@eugene-manuilov would you please add a QAB?

**Comment by eugene-manuilov** [comment]
Added.

**Comment by asvinb** [comment]
### QA: ✅ 

AdSense stories load fine and graphs display properly.

**Comment by aaemnnosttv** [comment]
## QA ❌ 

VRT is failing now on develop, the mock data seems to be different now, likely due to a missing fixed reference date. Assigning to @danielgent for a follow-up.

**Comment by danielgent** [comment]
I've created another PR to fix this issue

**Comment by aaemnnosttv** [comment]
Will check this again tomorrow but should be all set now 👍

**Comment by tofumatt** [comment]
AdSense stories and graphs look good and VRTs are passing pretty consistently regardless of dates now.

QA ✅
